Cleanup (#57)
sanger-tol/blobtoolkit: CHANGELOG.md

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [[1.0.0](https://github.com/sanger-tol/blobtoolkit/releases/tag/1.0.0)] – ??? - [2023-04-06]

### Enhancements & fixes

> Initial release of sanger-tol/blobtoolkit :tada

Reviewer: ```suggestion
Initial release of sanger-tol/blobtoolkit :tada:
```

Reviewer: You should also list somewhere the major differences with the Snakemake version, so that people don't expect a drop-in replacement, incl.
- Reads have to be aligned separately
- Two sub-workflows are still in progress
- The yaml config file is not used to control the actual pipeline
- The yaml config file does not reflect the actual software versions used by the pipeline, and may have other discrepancies

This release marks the point where the pipeline was moved from Snakemake at [blobtoolkit/blobtoolkit](https://github.com/blobtoolkit/blobtoolkit) over to Nextflow DSL2 at [sanger-tol/blobtoolkit](https://github.com/sanger-tol/blobtoolkit).

### Parameters

| Old parameter | New parameter |
| ------------- | ------------- |
|create_blobdir (blobtools replace): --evalue 1.0e-25 --hit-count 10 |  --evalue 1.0e-25 --hit-count 10 |
| diamond_blastp: --evalue 1.0e-25  --max-target-seqs 10 --max-hsps 1 | --evalue 1.0e-25  --max-target-seqs 10 --max-hsps 1 |
| diamond_blastx: --evalue 1.0e-25  --max-target-seqs 10              |               |

> | stats_windows: --window 0.1 --window 0.01 --window 100000 --window 1000000  |--window 0.1 --window 0.01 --window 100000 --window 1000000|

Reviewer: I don't understand the difference between the two columns: these are the same values, right ?

Reviewer 2: Yes, I was not sure if we should specify these parameters in the case they are the same in Snakemake and Nextflow versions. So, if they are the same they should not be included in the table, right?

Reviewer: > Yes, I was not sure if we should specify these parameters in the case they are the same in Snakemake and Nextflow versions. So, if they are the same they should not be included in the table, right?

Correct 👍🏼

> is present. </br> **NB:** Parameter has been **removed** if new parameter information isn't present.

### Software dependencies

Note, since the pipeline is using Nextflow DSL2, each process will be run with its own [Biocontainer](https://biocontainers.pro/#/registry). This means that on occasion it is e
ntirely possible for the pipeline to be using different versions of the same tool. However, the overall software dependency changes compared to the last release have been liste
d below for reference.

| Dependency | Old version | New version |
| ---------- | ----------- | ----------- |
|blastn      | 2.12.0+     |             |
|blobtools   | 4.0.7       |4.1.2        |
|busco       | 5.3.2       |5.4.3        |
|diamond     | 2.0.15      |2.0.15       |
|fasta_windows |           | 0.2.4       |
|goat-cli    | 0.2.0       |0.2.0        |
|gunzip      |             |1.10         |
|minimap2    | 2.24-r1122  |             |
|mosdepth    | 0.3.3       |0.3.3        |
|multiqc     |             |1.13         |
|python      | 3.9.13      |3.9.13       |
|samtools    | 1.15.1      |1.15.1       |

> |seqtk       | 1.3-r106    |             |

Reviewer 2: Ok. If the dependency version did not change between Snakemake and Nextflow versions, they should be removed, right?

> |seqtk       | 1.3-r106    |             |

Reviewer: Ah yes, I had missed that. If there isn't a change, it doesn't need to be reported

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [[1.0.0](https://github.com/sanger-tol/blobtoolkit/releases/tag/1.0.0)] – ??? - [2023-04-06]

### Enhancements & fixes

Initial release of sanger-tol/blobtoolkit :tada:

> - Required input: (1) Tree of Life organism ID (ToLID, --input) or (2) FASTA genome file (--fasta) and `samplesheet.csv` (--input) containing paths to aligned reads in CRAM format, see description of _Full samplesheet_ in `docs/usage.md`.

Reviewer: ```suggestion
- Required input: (1) Tree of Life organism ID (ToLID, `--input`) or (2) FASTA genome file (`--fasta`) and `samplesheet.csv` (`--input`) containing paths to aligned reads in CRAM format, see description of _Full samplesheet_ in `docs/usage.md`.
```

> - Lineages to run BUSCO are obtained using [goat_taxonsearch module](https://nf-co.re/modules/goat_taxonsearch) instead of reading them from `config.yaml` file (--taxon: NCBI taxonomy ID or species binomial name).

Reviewer: ```suggestion
- Lineages to run BUSCO are obtained using the [Genomes on a Tree (GoaT) API](https://goat.genomehubs.org/api) from the `--taxon` parameter (NCBI taxonomy ID or species binomial name) instead of reading them from `config.yaml` file.
```

Reviewer: I thought you said there was an option to provide the lineages in a file ?

Author: Yes `taxa_file`: https://github.com/sanger-tol/blobtoolkit/blob/dev/modules/nf-core/goat/taxonsearch/main.nf
I am going to test this a bit more with all the new changes and see how it is working.

- Versions of tools are collected using [custom_dumpsoftwareversions module](https://nf-co.re/modules/custom_dumpsoftwareversions) instead of reading them from `config.yaml` file.
- The pipeline requires aligned reads, `minimap2` is no longer included.

> - A GCA accession ID or YAML file is required (--accession or --yaml) to allow compatibility with `blobtoolkit` scripts. These are used to either generate or read a `config.yaml` file but do not pass parameters to the pipeline. Also software versions in this file do not reflect the actual versions used in the pipeline. 

Reviewer: Merge the first part (about the GCA accession) with the first item of the list ("required input")

Author: I am thinking making `accession` required even for draft. It can be used to define `meta`. The `yaml` will remain optional. I will only do this if it is a quick change.

- [fastawindows module](https://nf-co.re/modules/fastawindows) computes statistics across windows over the genome FASTA file.

> - Parameters for [diamond_blastp module](https://nf-co.re/modules/diamond_blastp) should be specified in `nextflow.config` file.

Reviewer: I would like to the underlying software, not the nf-core module

Reviewer 2: > I would like to the underlying software, not the nf-core module

Did you mean to use the actual link to the [fastawindows GitHub repo](https://github.com/tolkit/fasta_windows) instead of the link to the nf-core module?

Reviewer: Yes

This release marks the point where the pipeline was moved from Snakemake at [blobtoolkit/blobtoolkit](https://github.com/blobtoolkit/blobtoolkit) over to Nextflow DSL2 at [sanger-tol/blobtoolkit](https://github.com/sanger-tol/blobtoolkit). There are two subworkflows in the Snakemake version that are still being implemented in Nextflow: `diamond_blastx` and `blastn`.

### Parameters

| Old parameter | New parameter   |
| ------------- | --------------- |
|               | --accession     |
|               | --blastp_outext |
|               | --blastp_cols   |
|               | --fasta         |
|               | --input         |
|               | --taxon         |

> |               | --yaml          |

Reviewer: This table means that *every* parameter of the Snakemake pipeline is still supported and only new options were added, which I find strange.

Reviewer: For almost all our other pipelines, there isn't a previous version to compare against and the 1.0 entry of the changelog doesn't need a table "old parameter" vs "new parameter": `usage.md` suffices.
I'm OK doing the same here and skipping the table. However, there should still be somewhere in this file an introduction about BTK Nextflow 1.0, and how there isn't feature parity with the Snakemake version. We should also say:

- it will _never_ reach feature-parity because we're taking the occasion to reimplement things differently, c.f. the inclusion of `fasta_windows`
- the actual list of the BTK analyses implemented

Reviewer 2: > For almost all our other pipelines, there isn't a previous version to compare against and the 1.0 entry of the changelog doesn't need a table "old parameter" vs "new parameter": `usage.md` suffices. I'm OK doing the same here and skipping the table. However, there should still be somewhere in this file an introduction about BTK Nextflow 1.0, and how there isn't feature parity with the Snakemake version. We should also say:
> 
>     * it will _never_ reach feature-parity because we're taking the occasion to reimplement things differently, c.f. the inclusion of `fasta_windows`
> 
>     * the actual list of the BTK analyses implemented

@priyanka-surana @muffato Regarding this explanation about feature parity: should they be included in `usage.md `as well as `CHANGELOG.md`? Or at least a more detailed description on how Nextflow handles the parameters vs the Snakemake pipeline.

Author: I feel the best thing is to pretend our users know nothing about the snakemake pipeline, treat this as an independent pipeline. We can acknowledge and explain the relationship in README.md, CHANGELOG.md and CITATIONS.md, as appropriate but not too detailed.

I think both the parameters and the dependency table are useful for record keeping, but instead of comparing to snakemake, just add everything in the new column as this is first release. It will help users to have the record when we do the next release.

Reviewer: I don't think we can "pretend our users know nothing about the snakemake pipeline". We know that ToLA, Mark's team, Mara's team, etc, regularly use the Snakemake pipeline, and many of them can be considered advanced users. They need to know clearly, in advance, how the Nextflow version differs from the Snakemake one, otherwise they'll feel duped.

Reviewer 2: In the meantime I added a small description on how to specify parameters/inputs in Nextflow version compared to Snakemake version (also linked in CHANGELOG.md): 

https://github.com/sanger-tol/blobtoolkit/blob/cleanup/docs/usage.md#yaml-file-and-nextflow-configuration

This could be further improved by suggestions from the users as well. A couple of examples on how to change parameters could also be added in the documentation. Maybe if this part of the documentation is clear there is no need for the Parameters table, at least for the first release.

Author: > I don't think we can "pretend our users know nothing about the snakemake pipeline". We know that ToLA, Mark's team, Mara's team, etc, regularly use the Snakemake pipeline, and many of them can be considered advanced users. They need to know clearly, in advance, how the Nextflow version differs from the Snakemake one, otherwise they'll feel duped.

None of those users will be duped as they are aware of our work, plus duping people or taking undue credit has never been the intention. But I still prefer ease of understanding and a degree of abstraction in the documentation instead of over complicating with too much details. 

And as I mentioned before, a statement in  README.md, CHANGELOG.md and CITATIONS.md, should clarify the relationship without a deep analysis of the two versions. If the users still need further details, we can add a separate changelog - "CHANGELOG_snakemake.md" to track the differences with the snakemake version. However, if the main CHANGELOG shows differences with the snakemake version, how do we track differences between Nextflow versions in the future.

Happy to have this conversation off GitHub and clarify my points.

Author: @alxndrdiaz, Updates from offline conversation: We are going to write all the documentation for Nextflow independently. Then we will create a new section in usage that will detail the differences between the Snakemake and Nextflow versions (maybe something similar to the dependencies table) and will also provide instructions for users moving from Snakemake. This section will be referenced in both the README and CHANGELOG.

> **NB:** Parameter has been **updated** if both old and new parameter information is present. </br> **NB:** Parameter has been **added** if just the new parameter information
> is present. </br> **NB:** Parameter has been **removed** if new parameter information isn't present.

### Software dependencies

Note, since the pipeline is using Nextflow DSL2, each process will be run with its own [Biocontainer](https://biocontainers.pro/#/registry). This means that on occasion it is e
ntirely possible for the pipeline to be using different versions of the same tool. However, the overall software dependency changes compared to the last release have been liste

> d below for reference. Only `Docker` or `Singularity` containers are supported, `conda` is not supported.

Reviewer: The line breaks are funny

| Dependency             | Old version | New version |
| ---------------------- | ----------- | ----------- |
| blastn                 | 2.12.0      |             |
| blobtools              | 4.0.7       | 4.1.2       |
| busco                  | 5.3.2       | 5.4.3       |
| fasta_windows          |             | 0.2.4       |
| gunzip                 |             | 1.10        |
| minimap2               | 2.24-r1122  |             |
| multiqc                |             | 1.13        |
| seqtk                  | 1.3-r106    |             |
| windowmasker (blast\*) | 2.12.0      |             |

> **NB:** Dependency has been **updated** if both old and new version information is present. </br> **NB:** Dependency has been **added** if just the new version information is
> present. </br> **NB:** Dependency has been **removed** if version information isn't present.

> - blastn sub-workflow will be included in the next release.

Reviewer: Move this to the first list of the changelog. This is important in terms of functionality, not just software dependency

> - blast\* version that runs windowmasker is specified, windowmasker is no longer used in this version.

Reviewer: Is it because the new version of the pipeline expects a soft-masked assembly, whereas the Snakemake pipeline could do the masking itself ? That should be listed in the main list too.

Author: The snakemake version had windowmasker, we could bring it back in v2.0. The BLAST step is the one that really needs masked genome.
